enhancements/dns/externaldns-operator-aws-assume-role.md

```diff
@@ -0,0 +1,273 @@
+mentioned above differs from the design documented in the [API: DNS](/enhancements/installer/aws-cross-account-dns-zone.md#API-DNS)
+section of the AWS Shared VPC with Cross-account DNS Zones enhancement, which employs CRD validation utilizing a regular
+expression. Using the existing `IsARN` function provides more flexibility and reliability since it is maintained by AWS
+themselves, which helps to mitigate potential issues with edge cases.
+
+```go
+type ExternalDNSAWSProviderOptions struct {
+    [...]
+    // assumeRole is a reference to the IAM role that
+    // ExternalDNS will be assuming in order to perform
+    // any DNS updates.
+    //
+    // +kubebuilder:validation:Optional
+    // +optional
+    AssumeRole *ExternalDNSAWSAssumeRoleOptions `json:"assumeRole,omitempty"`
+}
+
+type ExternalDNSAWSAssumeRoleOptions struct {
+    // arn is an AWS role ARN that the ExternalDNS
+    // operator will assume when making DNS updates.
+    //
+    // +kubebuilder:validation:Required
+    // +required
+    ARN string `json:"arn,omitempty"`
+}
+
+```
+
+### Implementation Details/Notes/Constraints
+
+#### The Required Credentials Issue
+
+Extending the v1beta1 `ExternalDNS` API object revealed a pre-existing issue with the `Credentials` API field in
+`ExternalDNSAWSProviderOptions`. The problem is that `Credentials` is designated as required to satisfy vanilla
+Kubernetes use cases. On OpenShift, the ExternalDNS Operator generates a `CredentialRequest` automatically, eliminating
+the need to manually specify `Credentials`. This wasn't yet exposed as an issue because there were no other fields in
+`spec.provider.aws`, and `spec.provider.aws` is optional. The underlying [logic](https://github.com/openshift/external-dns-operator/blob/11e3b72b75d696b7419bce8360183373fc0725e1/api/v1beta1/externaldns_webhook.go#L132)
+of the webhook is accurate; the issue lies with this required field in the `ExternalDNS` CRD.
+
+However, there exists a simple workaround: OpenShift users can specify an empty `""` `Credentials` field as follows:
+```yaml
+apiVersion: externaldns.olm.openshift.io/v1beta1
+kind: ExternalDNS
+metadata:
+  name: sample-aws
+spec:
+  provider:
+    type: AWS
+    aws:
+      assumeRole:
+        arn: "arn:aws:xxx"
+      credentials:
+        name: ""
+```
+This satisfies the `ExternalDNS` CRD `+required` field while the ExternalDNS Operator logic [ignores](https://github.com/openshift/external-dns-operator/blob/bac092e98fe5f9065b75bd0fa21d5aeff9d00853/pkg/operator/controller/credentials-secret/controller.go#L259-L267)
+the empty string and uses the `CredentialsRequest`. As a result, we plan to introduce this `AssumeRole` field in the
+v1beta1 release accompanied by documentation that specifies this requirement to specify an empty `Credentials` field in
+OpenShift.
+
+In the next version of ExternalDNS Operator's API, we will address the requirement for the `Credentials` field as
```

Comment:
Unless users might be filling this with a real value now, I think it might be a better idea to start with the issue fixed and transparent to the user. Maybe write the code that fills credentials automatically when needed (or default to "" now and change the default later if needed?).

Reply:
I've been looking into what we could do to fill it with a real or default value.

I have three ideas I've explored:
1. Add `// +kubebuilder:default:={"name":""}` to the `Credentials` field v1beta1 CRD so it auto fills the default value
    - This works, but might break the upgrade path as laid out by the API team. I don't know if we can implement a CRD default, then convert to a pointer according to [backward-compatibility-gotchas correctly](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#backward-compatibility-gotchas). I'm checking with the API team.
2. Add code in ExternalDNS reconcile loop to add the default values when appropraite
    - This doesn't work because the reconcile loop happens after all validation (webooks, CRD validation  The CR can't be created then updated because it needs to pass validation.
3. Add a mutating webhook to default credentials to `""` when appropriate.
    - This is essentially the same outcome as # 1, but much more painful to implement, we don't have a mutating webhook and would need to build one just for this.

I'm following up with Andrey and the API team based on this suggestion. It's a valid concern, but a complicated one that'd we'd need concurrence from the API team to ensure we don't make our lives unnecessarily harder in the future. Hence, just telling the user to provide an empty credentials seemed the simplest answer,  but I agree that it's not the best UX.

Reference sequence for API validation: 
![image](https://github.com/openshift/enhancements/assets/4316443/bd2a831d-4bdf-4ec1-ba2a-3b054fcdfa35)

Reply:
**Update**: I finally got a [response](https://redhat-internal.slack.com/archives/CE4L0F143/p1695164515386059?thread_ts=1690552114.405509&cid=CE4L0F143) from the API team. They OK-ed the idea of adding a CRD default of `""` for the `Credentials` field, so this is a major improvement for user experience, making it seamless to the user.

Good idea to pushing for this, it was worth an ask from the API team, they seemed pleasantly shocked by the idea. The implementation update for defaulting is in this PR: https://github.com/openshift/external-dns-operator/pull/199. 

Enhancement Compare Link: https://github.com/openshift/enhancements/compare/b6c090e21f8a653c44c4f5c35faa1f5cf9af61e4..0a508b1baf394bf8fdb45efe2b030f0e4399fbd1